engine/graphics/src/opengl/graphics_opengl.cpp

```diff
@@ -2320,23 +2357,222 @@ static void LogFrameBufferError(GLenum status)
+        }
+        if (params.m_Namespace)
+        {
+            namespace_length = strlen(params.m_Namespace);
+            increase_bytes += namespace_length + 1;
+        }
+        if (params.m_InstanceName)
+        {
+            instance_name_length = strlen(params.m_InstanceName);
+            increase_bytes += instance_name_length + 1;
+        }
+
+        context->m_Buffer.OffsetCapacity(increase_bytes);
+        uint32_t buffer_offset = context->m_Buffer.Size();
+
+        char* write_ptr   = context->m_Buffer.Begin() + buffer_offset;
+        char* write_begin = write_ptr;
+
+        if (params.m_CanonicalName)
+        {
+            memcpy(write_ptr, params.m_CanonicalName, canonical_name_length);
+            path_pointers.m_OffsetCanonicalName     = buffer_offset + write_ptr - write_begin;
+            path_pointers.m_OffsetCanonicalNameUsed = 1;
+            write_ptr[canonical_name_length]        = 0;
+            write_ptr                              += canonical_name_length + 1;
+        }
+        if (params.m_Namespace)
+        {
+            memcpy(write_ptr, params.m_Namespace, namespace_length);
+            path_pointers.m_OffsetNamespace     = buffer_offset + write_ptr - write_begin;
+            path_pointers.m_OffsetNamespaceUsed = 1;
+            write_ptr[namespace_length]         = 0;
+            write_ptr                          += namespace_length + 1;
+        }
+        if (params.m_InstanceName)
+        {
+            memcpy(write_ptr, params.m_InstanceName, instance_name_length);
+            path_pointers.m_OffsetInstanceName     = buffer_offset + write_ptr - write_begin;
+            path_pointers.m_OffsetInstanceNameUsed = 1;
+            write_ptr[instance_name_length]        = 0;
+            write_ptr                             += instance_name_length + 1;
+        }
+
+        context->m_Buffer.SetSize(context->m_Buffer.Capacity());
+
+        context->m_Paths.OffsetCapacity(1);
+        context->m_Paths.Push(path_pointers);
+    }
+
+    // Create a canonical path for all leaf members in uniform buffers, e.g:
+    //
+    // Case 1:
+    // struct PBRMaterial
+    // {
+    //     float metallic;
+    //     float roughness;
+    //     ...
+    // };
+    // uniform fs_uniforms {
+    //     PBRMaterial material;
+    // };
+    // member 0: fs_uniforms.material.metallic;
+    // member 1: fs_uniforms.material.roughness;
+    //
+    // Case 2:
+    // We have the same, but we also use an "instance member" to rename the uniform (will also be useful later for arrayed uniforms):
+    // uniform fs_uniforms {
+    //     PBRMaterial material;
+    // } my_material;
+    // member 0: my_material.material.metallic;
+    // member 1: my_material.material.roughness;
+    static char* GetConstructedCanonicalName(BuildCanonicalPathsCallbackContext& context, char* str, uint32_t str_len, char* canonical_name_buffer, uint32_t canonical_name_buffer_len)
```

Comment:
Not sure about doing all of this work in runtime, but even if we had all the prepared paths in the data formats we still need to do this "fixup" of the paths because of how OpenGL returns the reflection data. It's only really needed for OpenGL, so I don't know if we want to do more work in the pipeline to avoid this?

Ideally I'd like to use the reflection as-is for OpenGL, but unfortunately we still need to generate the OpenGL reflection data to get the bind locations so we can bind resource to the correct shader bind points. I don't think we want OpenGL to be driving our data formats even further, since it's becoming more and more the odd one out.